packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +959,1336 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+    }
+    assert(_sortedChildren.contains(child));
+
+    if (child._overlayChildRenderBox != null) {
+      _theatre.markNeedsPaint();
+      _theatre.markNeedsCompositingBitsUpdate();
+      _theatre.markNeedsSemanticsUpdate();
+    }
+  }
+
+  void _remove(_OverlayEntryLocation child) {
+    final bool wasInCollection = _sortedChildren.remove(child);
+    assert(wasInCollection);
+    if (child._overlayChildRenderBox != null) {
+      _theatre.markNeedsPaint();
+      _theatre.markNeedsCompositingBitsUpdate();
+      _theatre.markNeedsSemanticsUpdate();
+    }
+  }
+
+  _OverlayEntryLocation? _childAfter(_OverlayEntryLocation? child) {
+    if (_sortedChildren.isEmpty) {
+      return null;
+    }
+    final _OverlayEntryLocation? candidate = child == null ? _sortedChildren.first : child.next;
+    return candidate != null ? candidate._overlayChildRenderBox != null ? candidate : _childAfter(candidate) : null;
+  }
+
+  _OverlayEntryLocation? _childBefore(_OverlayEntryLocation? child) {
+    if (_sortedChildren.isEmpty) {
+      return null;
+    }
+    final _OverlayEntryLocation? candidate = child == null ? _sortedChildren.last : child.previous;
+    return candidate != null ? candidate._overlayChildRenderBox != null ? candidate : _childBefore(candidate) : null;
+  }
+
+  static _ChildIterator? _mapIterator(_Iterator<_OverlayEntryLocation>? fromIterator) {
+    if (fromIterator == null) {
+      return null;
+    }
+
+    _RenderDeferredLayoutBox? iter() {
+      final _OverlayEntryLocation? location = fromIterator();
+      if (location == null) {
+        return null;
+      }
+      return location._overlayChildRenderBox ?? iter();
+    }
+
+    return iter;
+  }
+
+  _Iterator<_OverlayEntryLocation>? _forwardIterator() {
+    // Since the _Iterator returned by this method will be used immediately,
+    // it's OK to evaluate the first value and determine if we can just return
+    // null.
+    _OverlayEntryLocation? child = _childAfter(null);
+    if (child == null) {
+      return null;
+    }
+    bool isInitial = true;
+    return () {
+      if (isInitial) {
+        assert(child != null);
+        assert(_sortedChildren.contains(child));
+        isInitial = false;
+        return child;
+      }
+      return child == null ? null : child = _childAfter(child);
+    };
+  }
+
+  _Iterator<_OverlayEntryLocation>? _backwardIterator() {
+    // Since the _Iterator returned by this method will be used immediately,
+    // it's OK to evaluate the first value and determine if we can just return
+    // null.
+    _OverlayEntryLocation? child = _childBefore(null);
+    if (child == null) {
+      return null;
+    }
+    bool isInitial = true;
+    return () {
+      if (isInitial) {
+        assert(child != null);
+        assert(_sortedChildren.contains(child));
+        isInitial = false;
+        return child;
+      }
+      return child == null ? null : child = _childBefore(child);
+    };
+  }
+
+  @override
```

Comment:
Haven't looked at this too closely, but why is the iterator business so complex instead of just doing _sortedChildren.iterator?

Reply:
that iterator will throw a concurrent modification error if used in performLayout I believe, and some elements need to be filtered out.